### PR TITLE
Need to also emit legacy state for read override connectors when state is not specified

### PR DESCRIPTION
**Pull request by brianjlai**
## What
When fixing https://github.com/airbytehq/airbyte/pull/16566 , I forgot to also emit the legacy format when state isn't specified.

## How
Since we may need to conditionally emit the legacy format in more than one place, I refactored the condition into it's own function so it could be reused.

Tested again w/ the following conditions work with the correct state interpretation:

- No `--state` specified in the read request: replicate all records, emit final state
- state file is `{}`: replicate all records, emit final state
- state file is`[]`: replicate all records, emit final state
- state file contains valid state: emit no records, emit final state

### CODE REVIEW

File: airbyte-cdk/python/airbyte_cdk/sources/source.py [hunk 1]
```diff
@@ -66,13 +67,23 @@ def read_state(self, state_path: str) -> List[AirbyteStateMessage]:
...
+
+    def _emit_legacy_state_format(self, state_obj) -> Union[List[AirbyteStateMessage], MutableMapping[str, Any]]:
+        """
+        Existing connectors that override read() might not be able to interpret the new state format. We temporarily
+        send state in the old format for these connectors, but once all have been upgraded, this method can be removed,
+        and we can then emit state in the list format.
+        """
+        # vars(self.__class__) checks if the current class directly overrides the read() function
+        if "read" in vars(self.__class__):
+            return defaultdict(dict, state_obj)
```
Comment (alafanechere) [inline airbyte-cdk/python/airbyte_cdk/sources/source.py:81]: Could you share why you call `defaultdict` with a `dict` default factory?

**Review by evantahler — APPROVED** [review]
👍 from me, but deferring to @alafanechere's review

File: airbyte-cdk/python/airbyte_cdk/sources/source.py (hunk 1 above)
Comment (brianjlai) [reply airbyte-cdk/python/airbyte_cdk/sources/source.py:81]: Great question! Not sure if it actually makes a difference, but the original code before the changes also wrapped it in `defaultdict` so I wanted to preserve the same functionality. Reference to prior changes: https://github.com/airbytehq/airbyte/blob/055763ec73e2dd260715b8b108664cf96dae5d23/airbyte-cdk/python/airbyte_cdk/sources/source.py#L49

**Review by alafanechere — APPROVED** [review]
Lgtm!

**Pull request merged**